stan/math/rev/core/var.hpp

```diff
@@ -322,7 +566,7 @@ class var_value {
    * @param num_cols Number of columns to return.
    */
   inline auto block(Eigen::Index start_row, Eigen::Index start_col,
-                    Eigen::Index num_rows, Eigen::Index num_cols) const {
+                    Eigen::Index num_rows, Eigen::Index num_cols) {
```

Comment:
Should they be const? Like if we do `A.block(0, 0, 2, 2)` that is returning back a new type, but that type will modify the underlying `vari*` that was returned from the const method which idt is supposed to happen right?

Reply:
So far in the whole math library const var or const matrix of vars means only that the var or matrix of var is const, but not the underlying data. Whole library is modifying adjoints of vars and matrices it got in as const arguments.

The problem will be that most of times matrices are passed around by const references. So if this function is not const, a function accepting a matrix will not be able to get a block of its argument.

Alternatively you will need to duplicate all these methods to have const aversion returning a const object and a non-const version returning a non-const object.

Reply:
I ended up going with two versions (const and non-const). The code duplication is a bummer but when I tried making all of these just const I ended up with some const-correctness errors from the compiler that were kind of complicated and I couldn't sort out. I also tried a scheme like in the stackoverflow post below but couldn't sort it out because we have multiple levels of objects and we'd need to do some scheme where we make the inner types from the const version also non-const

https://stackoverflow.com/a/47369227/2269255

Reply:
If you have a way to do this without doing the duplication I'm v open to it! I just couldn't figure it out lol

Reply:
From the top of my head no, but I might try to deduplicate it next time I touch this code.

Reply:
Actually it looks like const and non const overloads have exactly the same code. Which makes the non-const overloads redundant.

Reply:
You can leave both overloads of the `eval()`, as the return type actually differs.

Reply:
I think they need to be separate because of the eigen types that are returned from the const and non-const versions. Like if I leave only the const versions then I'm seeing errors where the const returned eigen type can't be cast to the non-const eigen type (which makes sense).

```
./stan/math/rev/core/var.hpp: In instantiation of ‘auto stan::math::var_value<T, typename std::enable_if<std::integral_constant<bool, (stan::is_eigen<T>::value || stan::is_matrix_cl<T>::value)>::value, void>::type>::block(Eigen::Index, Eigen::Index, Eigen::Index, Eigen::Index) const [with T = Eigen::Matrix<double, -1, -1>; typename std::enable_if<std::integral_constant<bool, (stan::is_eigen<T>::value || stan::is_matrix_cl<T>::value)>::value, void>::type = void; Eigen::Index = long int]’:
test/unit/math/rev/core/var_test.cpp:181:38:   required from here
./stan/math/rev/core/var.hpp:565:12: error: no matching function for call to ‘stan::math::var_value<Eigen::Block<const Eigen::Map<Eigen::Matrix<double, -1, -1> >, -1, -1, false>, void>::var_value(vari_sub*)’
  565 |     return var_sub(
      |            ^~~~~~~~
  566 |         new vari_sub(vi_->block(start_row, start_col, num_rows, num_cols)));
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./stan/math/rev/core/var.hpp:397:3: note: candidate: ‘stan::math::var_value<T, typename std::enable_if<std::integral_constant<bool, (stan::is_eigen<T>::value || stan::is_matrix_cl<T>::value)>::value, void>::type>::var_value(stan::math::var_value<T, typename std::enable_if<std::integral_constant<bool, (stan::is_eigen<T>::value || stan::is_matrix_cl<T>::value)>::value, void>::type>::vari_type*) [with T = Eigen::Block<const Eigen::Map<Eigen::Matrix<double, -1, -1> >, -1, -1, false>; typename std::enable_if<std::integral_constant<bool, (stan::is_eigen<T>::value || stan::is_matrix_cl<T>::value)>::value, void>::type = void; stan::math::var_value<T, typename std::enable_if<std::integral_constant<bool, (stan::is_eigen<T>::value || stan::is_matrix_cl<T>::value)>::value, void>::type>::vari_type = stan::math::vari_view<Eigen::Block<const Eigen::Map<Eigen::Matrix<double, -1, -1> >, -1, -1, false>, void>]’
  397 |   var_value(vari_type* vi) : vi_(vi) {}  // NOLINT
      |   ^~~~~~~~~
./stan/math/rev/core/var.hpp:397:24: note:   no known conversion for argument 1 from ‘vari_sub*’ {aka ‘stan::math::vari_view<const Eigen::Block<const Eigen::Map<Eigen::Matrix<double, -1, -1> >, -1, -1, false>, void>*’} to ‘stan::math::var_value<Eigen::Block<const Eigen::Map<Eigen::Matrix<double, -1, -1> >, -1, -1, false>, void>::vari_type*’ {aka ‘stan::math::vari_view<Eigen::Block<const Eigen::Map<Eigen::Matrix<double, -1, -1> >, -1, -1, false>, void>*’}
  397 |   var_value(vari_type* vi) : vi_(vi) {}  // NOLINT
      |             ~~~~~~~~~~~^~
```

Reply:
I think they need to be separate because of the eigen types that are returned from the const and non-const versions. Like if I leave only the const versions then I'm seeing errors where the const returned eigen type can't be cast to the non-const eigen type (which makes sense).

```
./stan/math/rev/core/var.hpp: In instantiation of ‘auto stan::math::var_value<T, typename std::enable_if<std::integral_constant<bool, (stan::is_eigen<T>::value || stan::is_matrix_cl<T>::value)>::value, void>::type>::block(Eigen::Index, Eigen::Index, Eigen::Index, Eigen::Index) const [with T = Eigen::Matrix<double, -1, -1>; typename std::enable_if<std::integral_constant<bool, (stan::is_eigen<T>::value || stan::is_matrix_cl<T>::value)>::value, void>::type = void; Eigen::Index = long int]’:
test/unit/math/rev/core/var_test.cpp:181:38:   required from here
./stan/math/rev/core/var.hpp:565:12: error: no matching function for call to ‘stan::math::var_value<Eigen::Block<const Eigen::Map<Eigen::Matrix<double, -1, -1> >, -1, -1, false>, void>::var_value(vari_sub*)’
  565 |     return var_sub(
      |            ^~~~~~~~
  566 |         new vari_sub(vi_->block(start_row, start_col, num_rows, num_cols)));
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./stan/math/rev/core/var.hpp:397:3: note: candidate: ‘stan::math::var_value<T, typename std::enable_if<std::integral_constant<bool, (stan::is_eigen<T>::value || stan::is_matrix_cl<T>::value)>::value, void>::type>::var_value(stan::math::var_value<T, typename std::enable_if<std::integral_constant<bool, (stan::is_eigen<T>::value || stan::is_matrix_cl<T>::value)>::value, void>::type>::vari_type*) [with T = Eigen::Block<const Eigen::Map<Eigen::Matrix<double, -1, -1> >, -1, -1, false>; typename std::enable_if<std::integral_constant<bool, (stan::is_eigen<T>::value || stan::is_matrix_cl<T>::value)>::value, void>::type = void; stan::math::var_value<T, typename std::enable_if<std::integral_constant<bool, (stan::is_eigen<T>::value || stan::is_matrix_cl<T>::value)>::value, void>::type>::vari_type = stan::math::vari_view<Eigen::Block<const Eigen::Map<Eigen::Matrix<double, -1, -1> >, -1, -1, false>, void>]’
  397 |   var_value(vari_type* vi) : vi_(vi) {}  // NOLINT
      |   ^~~~~~~~~
./stan/math/rev/core/var.hpp:397:24: note:   no known conversion for argument 1 from ‘vari_sub*’ {aka ‘stan::math::vari_view<const Eigen::Block<const Eigen::Map<Eigen::Matrix<double, -1, -1> >, -1, -1, false>, void>*’} to ‘stan::math::var_value<Eigen::Block<const Eigen::Map<Eigen::Matrix<double, -1, -1> >, -1, -1, false>, void>::vari_type*’ {aka ‘stan::math::vari_view<Eigen::Block<const Eigen::Map<Eigen::Matrix<double, -1, -1> >, -1, -1, false>, void>*’}
  397 |   var_value(vari_type* vi) : vi_(vi) {}  // NOLINT
      |             ~~~~~~~~~~~^~
```

Reply:
In var? For both I'm seeing the const and non-const version of each